mypy/semanal.py

```diff
@@ -2931,21 +2931,42 @@ def visit_member_expr(self, expr: MemberExpr) -> None:
                 if full_name in obsolete_name_mapping:
                     self.fail("Module%s has no attribute %r (it's now called %r)" % (
                         mod_name, expr.name, obsolete_name_mapping[full_name]), expr)
-        elif isinstance(base, RefExpr) and isinstance(base.node, TypeInfo):
-            n = base.node.names.get(expr.name)
-            if n is not None and (n.kind == MODULE_REF or isinstance(n.node, TypeInfo)):
-                # This branch handles the case C.bar where C is a class and
-                # bar is a type definition or a module resulting from
-                # `import bar` inside class C. Here base.node is a TypeInfo,
-                # and again we look up the name in its namespace.
-                # This is done only when bar is a module or a type; other
-                # things (e.g. methods) are handled by other code in checkmember.
-                n = self.normalize_type_alias(n, expr)
-                if not n:
-                    return
-                expr.kind = n.kind
-                expr.fullname = n.fullname
-                expr.node = n.node
+        elif isinstance(base, RefExpr):
+            # This branch handles the case C.bar (or cls.bar or self.bar inside
+            # a classmethod/method), where C is a class and bar is a type
+            # definition or a module resulting from `import bar` (or a module
+            # assignment) inside class C. We look up bar in the class' TypeInfo
+            # namespace.  This is done only when bar is a module or a type;
+            # other things (e.g. methods) are handled by other code in
+            # checkmember.
+            type_info = None
+            if isinstance(base.node, TypeInfo):
+                # C.bar where C is a class
+                type_info = base.node
+            elif isinstance(base.node, Var) and self.function_stack:
```

Comment:
Inexperience in the mypy codebase! I missed the existence of `self.type`, thanks! Pushing update.

This complex code is performing two tasks, though, and `self.type` addresses only one of them.  The other task is verifying that `base.node` is in fact a reference to the class (`cls`) or an instance of it (`self`). If we fully simplify, we fail this added test, because we will consider any variable as a class reference: https://gist.github.com/carljm/174bff5aeb78d9a3b04ed11a536cfb9f

If you have a suggestion of a better way to do this verification, I'd be happy for that! Of course it is possible to just check that the variable name is `cls` or `self`, but this is not really correct; there is no such naming requirement in the language, and in some nested-class cases I've seen people use e.g. `self_`.